runtime/common/KernelWrapper.h

```diff
@@ -0,0 +1,347 @@
+      : m_buffer(buffer), m_remaining(remaining) {}
+  bool read(char *data, std::size_t size) {
+    if (size > m_remaining)
+      return false;
+    std::memcpy(data, m_buffer, size);
+    m_buffer += size;
+    m_remaining -= size;
+    return true;
+  }
+
+  const char *data() const { return m_buffer; }
+
+private:
+  const char *m_buffer = nullptr;
+  std::size_t m_remaining = 0;
+};
+
+//===----------------------------------------------------------------------===//
+//
+// Utilities to package (serialize) kernel arguments as a flat byte buffer.
+// This is equivalent to the bridge-generated args creator function for each
+// kernel signature type.
+//
+//===----------------------------------------------------------------------===//
+
+// Specialize to describe how to serialize/deserialize to/from the given
+// concrete type.
+template <typename T, typename _ = void>
+class SerializeArgImpl;
+
+// A utility class for serializing kernel `args` as a flat buffer
+template <typename... ArgTs>
+class SerializeArgs;
+
+// Empty list specialization
+template <>
+class SerializeArgs<> {
+public:
+  static std::size_t size() { return 0; }
+  static bool serialize(SerializeOutputBuffer &) { return true; }
+  static bool deserialize(SerializeInputBuffer &) { return true; }
+};
+
+// Non-empty list specialization for SerializeArgs.
+template <typename ArgT, typename... ArgTs>
+class SerializeArgs<ArgT, ArgTs...> {
+public:
+  static std::size_t size(const ArgT &arg, const ArgTs &...args) {
+    return SerializeArgImpl<ArgT>::size(arg) +
+           SerializeArgs<ArgTs...>::size(args...);
+  }
+
+  static bool serialize(SerializeOutputBuffer &buf, const ArgT &arg,
+                        const ArgTs &...args) {
+    return SerializeArgImpl<ArgT>::serialize(buf, arg) &&
+           SerializeArgs<ArgTs...>::serialize(buf, args...);
+  }
+
+  static bool deserialize(SerializeInputBuffer &buf, ArgT &arg,
+                          ArgTs &...args) {
+    return SerializeArgImpl<ArgT>::deserialize(buf, arg) &&
+           SerializeArgs<ArgTs...>::deserialize(buf, args...);
+  }
+};
+
+// Serialization for 'trivial' types (POD)
+template <typename T>
+class SerializeArgImpl<T, std::enable_if_t<std::is_trivial<T>::value>> {
+public:
+  static std::size_t size(const T &x) { return sizeof(T); }
+
+  static bool serialize(SerializeOutputBuffer &buf, const T &x) {
+    T tempVal = x;
+    return buf.write(reinterpret_cast<const char *>(&tempVal), sizeof(tempVal));
+  }
+
+  static bool deserialize(SerializeInputBuffer &buf, T &x) {
+    T tempVal;
+    if (!buf.read(reinterpret_cast<char *>(&tempVal), sizeof(tempVal)))
+      return false;
+    x = tempVal;
+    return true;
+  }
+};
+
+// Serialization for a vector of 'trivial' types (POD), e.g.,
+// `std::vector<double>`.
+// Format: vector size followed by data
+// Note: vectors of non-POD types are not supported by this library-mode
+// serializer.
```

Comment:
Will this work for `vector<vector<... AT ...>>`? (`AT` is an arithmetic type.)

We'll be adding product types (struct-like objects) of `AT`, vectors of product types, and product types with vector components in the bridge as well.

Reply:
No, it won't work for nested vectors. I'll look into adding support for this adopting the bridge convention.

Reply:
I've added https://github.com/NVIDIA/cuda-quantum/issues/1088 to track this as a follow-up.